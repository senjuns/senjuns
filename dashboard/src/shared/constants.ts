import {
  MeasurementTypes,
  MeasurementStatisticsTypes,
  TPasswordRule,
} from '../shared/interfaces';

export const MeasurementUnit = {
  CelciusDegree: '°C',
  FahrenheitDegree: '°F',
  Percent: '%',
  KP: 'kP',
  mph: 'mph',
  ppm: 'ppm',
  DLI: 'DLI',
  μMol_m2_s: 'μMol/m2/s',
};

export const NotificationLevel = {
  Nominal: 'nominal',
  Warning: 'warning',
  Error: 'error',
};

export const ScreenSize = {
  xs: 576,
  sm: 768,
  md: 992,
  lg: 1200,
  xl: 1400,
};

export const Colors = {
  brightBlue100: '#60C3E3',
  brightBlue80: '#7FCEE8',
  brightBlue60: '#9FDAEE',
  grey100: '#E5E5E5',
  grey80: '#EAEAEA',
  grey60: '#EFEFEF',
  black: '#000000',
  dark1: '#383838',
  dark7: '#808080',
  dark8: '#B4B4B4',
  dark9: '#CFCFCF',
  green1: '#56C254',
  grey5: '#E0E0E0',
  hb1: '#82879A',
  hb2: '#9BA0B6',
  hb3: '#C3C7D9',
  hb4: '#ECEDF1',
  hb5: '#F0F1F4',
  hb6: '#F8F8FC',
  orange1: '#DD382F',
  orange2: '#E86339',
  orange3: '#F48E42',
  orange4: '#FFB94C',
  white: '#ffffff',
  linearGradientedOrange: 'linear-gradient(90deg, #E86339 0%, #FFB94C 100%)',
};

export const Sizes = {
  xxSmall: 4,
  xSmall: 8,
  small: 12,
  medium: 16,
  large: 20,
  xLarge: 24,
  xxLarge: 32,
};

export const Fonts = {
  h1: { size: 48, style: 'normal', weight: 600 },
  h2: { size: 40, style: 'normal', weight: 600 },
  h3: { size: 36, style: 'normal', weight: 600 },
  h4: { size: 24, style: 'normal', weight: 600 },
  h5: { size: 20, style: 'normal', weight: 600 },
  h6: { size: 18, style: 'normal', weight: 600 },
  body1: { size: 16, style: 'normal', weight: 'normal' },
  body2: { size: 14, style: 'normal', weight: 'normal' },
};

export const MEASUREMENT_OPTION_DETAILS = {
  [MeasurementTypes.Light]: {
    statisticsKey: MeasurementStatisticsTypes.AirTemperature,
    unit: MeasurementUnit.FahrenheitDegree,
  },
  [MeasurementTypes.Temperature]: {
    statisticsKey: MeasurementStatisticsTypes.AirTemperature,
    unit: MeasurementUnit.FahrenheitDegree,
  },
  [MeasurementTypes.Humidity]: {
    statisticsKey: MeasurementStatisticsTypes.RelativeHumidity,
    unit: MeasurementUnit.Percent,
  },
  [MeasurementTypes.LeafVpd]: {
    statisticsKey: MeasurementStatisticsTypes.LeafVpd,
    unit: MeasurementUnit.KP,
  },
  [MeasurementTypes.Vpd]: {
    statisticsKey: MeasurementStatisticsTypes.AirVpd,
    unit: MeasurementUnit.KP,
  },
  [MeasurementTypes.Airflow]: {
    statisticsKey: MeasurementStatisticsTypes.AirFlow,
    unit: MeasurementUnit.mph,
  },
  [MeasurementTypes.Co2]: {
    statisticsKey: MeasurementStatisticsTypes.AirCo2,
    unit: MeasurementUnit.ppm,
  },
  [MeasurementTypes.Ppfd]: {
    statisticsKey: MeasurementStatisticsTypes.Par,
    unit: MeasurementUnit.μMol_m2_s,
  },
  [MeasurementTypes.LeafTemperature]: {
    statisticsKey: MeasurementStatisticsTypes.LeafTemperature,
    unit: MeasurementUnit.FahrenheitDegree,
  },
};

export const APP_URL = {
  login: '/login',
  forgot: '/forgot',
  reset: '/reset',
  notFound: '/not-found',
  resetPassword: (username: string) => `/reset?username=${username}`,
  userHome: '/user/home',
  userAbout: '/user/about',
  home: '/',
  roomDetails: (id: number) => `/room-details/${id}`,
  roomDetailsOverview: (id: number) => `/room-details/${id}/overview`,
  roomDetailsHeatMap: (id: number) => `/room-details/${id}/heat-map`,
  roomDetailsLineChart: (id: number) => `/room-details/${id}/line-chart`,
  roomDetailsPhotoFeed: (id: number) => `/room-details/${id}/photo-feed`,
};

export enum ROOM_DETAIL_TABS {
  OVERVIEW = 'overview-tab',
  LINE_CHART = 'line-chart-tab',
  HEAT_MAP = 'heat-map-tab',
  PHOTO_FEED = 'photo-feed-tab',
}

export const AWS_API_CODE = {
  // will return when username or password incorrect or when user does not exists.
  user_not_found_exception: 'UserNotFoundException',
};

export const PASSWORD_RULES: TPasswordRule[] = [
  {
    label: 'Password must match',
    check: (password: string, confirmPassword: string): boolean =>
      password === confirmPassword,
  },
  {
    label: 'At least 8 characters',
    check: (password: string): boolean => password.length >= 8,
  },
  {
    label: 'Must contain at least one upper case letter',
    check: (password: string): boolean => /[A-Z]/.test(password),
  },
  {
    label: 'Must contain at least one special character',
    check: (password: string): boolean => /[#?!@$%^&*-]/.test(password),
  },
  {
    label: 'Must contain at least one number',
    check: (password: string): boolean => /[0-9]/.test(password),
  },
];

export const FEATURE_ENABLED = 'FEATURE_ENABLED';

export const NULL_VALUE = -1;

export const Z_INDEXES = {
  DROPDOWN: 10001,
};
