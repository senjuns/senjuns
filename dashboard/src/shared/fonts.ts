import { css } from 'styled-components';

export const InterBoldMirage16px = css`
  color: var(--mirage);
  font-family: var(--font-family-inter);
  font-size: var(--font-size-s);
  font-weight: 700;
  font-style: normal;
`;

export const Poppins18 = css`
  font-family: var(--font-family-poppins);
  font-size: var(--font-size-l2);
  letter-spacing: 0;
  font-weight: 400;
  font-style: normal;
`;

export const Poppins22 = css`
  font-family: var(--font-family-roboto);
  font-size: var(--font-size-22px);
  letter-spacing: 0;
  font-weight: 400;
  font-style: normal;
`;

export const Poppins26 = css`
  font-family: var(--font-family-poppins);
  font-size: var(--font-size-xxl2);
  letter-spacing: 0;
  font-weight: 400;
  font-style: normal;
`;

export const RobotoBoldBlack22px = css`
  color: var(--black);
  font-family: var(--font-family-roboto);
  font-size: var(--font-size-22px);
  font-weight: 700;
  font-style: normal;
`;
