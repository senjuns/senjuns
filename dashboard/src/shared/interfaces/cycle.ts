export interface IPlantInfo {
  strainName: string;
  count: number;
}
