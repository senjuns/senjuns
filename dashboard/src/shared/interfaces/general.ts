export interface Size {
  width: number;
  height: number;
}

export enum METRICS_TYPE {
  USER_LOGIN,
  USER_LOGOUT,
}

export enum METRICS_CATEGORY {
  USER = 'User',
}

export enum FEATURE_FLAGS {
  HEATMAP = 'FEATURE_FLAGS_HEATMAP',
  SETTINGS = 'FEATURE_FLAGS_SETTINGS',
  DASHBOARD = 'FEATURE_FLAGS_DASHBOARD',
  LINE_CHART = 'FEATURE_FLAGS_LINE_CHART',
  CALENDAR = 'FEATURE_FLAGS_LINE_CALENDAR',
  NOTIFICATION = 'FEATURE_FLAGS_NOTIFICATION',
  ANCHOR_DATE_TIME = 'FEATURE_FLAGS_ANCHOR_DATE_TIME',
  DEBUG = 'FEATURE_FLAGS_LOG',
  INTERNAL_HEATMAP_TOOL_FLAG = 'FEATURE_INTERNAL_HEATMAP_TOOL_FLAG',
  AIRFLOW_SENSOR_FLAG = 'AIRFLOW_SENSOR_FLAG',
}

export type TPasswordRule = {
  label: string;
  check: (password: string, confirmPassword: string) => boolean;
};

export type ArrayElement<ArrayType extends readonly unknown[]> =
  ArrayType extends readonly (infer ElementType)[] ? ElementType : never;

export type Nullable<T> = T | null;

export type Maybe<T> = T | null | undefined;

export interface IPosition {
  x: number;
  y: number;
}

export interface MetricsMeta {
  email?: string;
}

export interface TimeRange {
  start: Date;
  end: Date;
}
