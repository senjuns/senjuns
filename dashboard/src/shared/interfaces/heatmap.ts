export type HeatMapStylings = {
  low: {
    max: number;
    color: string;
  };
  mediumLow: {
    max: number;
    color: string;
  };
  medium: {
    max: number;
    color: string;
  };
  mediumHigh: {
    color: string;
    max: number;
  };
  high: {
    color: string;
    max: number;
  };
};
