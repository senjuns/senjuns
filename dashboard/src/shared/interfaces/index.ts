export * from './cycle';
export * from './general';
export * from './measurement';
export * from './section';
