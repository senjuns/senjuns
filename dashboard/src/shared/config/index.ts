import Cookies from 'js-cookie';
// import { FEATURE_ENABLED } from '../../shared/constants';
import { FEATURE_FLAGS } from '../../shared/interfaces';
// @ts-ignore Needs to be ignored as it is dynamically created.
import config from './config';

const defaultConfig: any = config;

if (Cookies.get(FEATURE_FLAGS.DEBUG) === 'FEATURE_ENABLED') {
  console.log(`process.env: ${JSON.stringify(process.env, null, 2)}`);
  console.log(`defaultConfig: ${JSON.stringify(defaultConfig, null, 2)}`);
}

export default defaultConfig;
