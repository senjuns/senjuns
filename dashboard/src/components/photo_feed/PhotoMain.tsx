import Konva from 'konva';
import { FC, useCallback, useEffect, useRef } from 'react';
import { Stage, Layer } from 'react-konva';

import SectionsLayer from '../../components/photo_feed/SectionsLayer';
import { Size, HistoryInformation } from '../../shared/interfaces';

interface PhotoMainProps {
  /**
   * Size of the images grid.
   */
  imageSize: Size;
  /**
   * The history information of the system.
   */
  historyInformation: HistoryInformation;
  /**
   * Current miniMap cursor position.
   */
  miniMapPosition: Konva.Vector2d;
  /**
   * Check if the scale is available.
   */
  checkIsOutOfScale: (scale: number) => boolean;
  /**
   * Callback to be called when the cursor position changes.
   */
  onChangePosition: (value: Konva.Vector2d) => void;
  /**
   * Callback to be called when the zoom scale changes.
   */
  onChangeScale: (value: number) => void;
  /**
   * Callback to be called when the zoom scale delta changes.
   */
  onChangeScaleDelta: (delta: number) => void;
  /**
   * Callback to be called when the section in x,y is clicked.
   */
  onClick: (x: number, y: number) => void;
  /**
   * The size of presentation layer to be displayed in a window.
   */
  presentationSize: Size;
  /**
   * The current scale.
   */
  scale: number;
  /**
   * The scale delta that is updated outside of this component.
   * ie) Zoom In/Out clicking -/+ will update this prop to control scale.
   */
  scaleDelta: number;
}

const PhotoMain: FC<PhotoMainProps> = ({
  imageSize,
  historyInformation,
  miniMapPosition,
  checkIsOutOfScale,
  onChangePosition,
  onChangeScale,
  onChangeScaleDelta,
  onClick,
  presentationSize,
  scale,
  scaleDelta,
}) => {
  const stageRef = useRef<Konva.Stage>(null);

  const changePosition = (x: number, y: number) => {
    const stage = stageRef.current;
    if (!stage) return null;

    // eslint-disable-next-line @typescript-eslint/no-shadow
    const scale = stage.scaleX();
    const newPosition = {
      x: Math.max(
        Math.min(0, x),
        presentationSize.width - imageSize.width * scale
      ),
      y: Math.max(
        Math.min(0, y),
        presentationSize.height - imageSize.height * scale
      ),
    };

    stage.position(newPosition);
    stage.batchDraw();
    onChangePosition(newPosition);
    return null;
  };

  const handleChangeScaleDelta = useCallback(() => {
    if (!scaleDelta || !stageRef.current) {
      return;
    }

    const newScale = scale * scaleDelta;

    const stage = stageRef.current;
    const center = {
      x: stage.width() / 2,
      y: stage.height() / 2,
    };
    const relatedTo = {
      x: (center.x - stage.x()) / scale,
      y: (center.y - stage.y()) / scale,
    };

    stage.scale({ x: newScale, y: newScale });
    changePosition(
      center.x - relatedTo.x * newScale,
      center.y - relatedTo.y * newScale
    );
    onChangeScale(newScale); // needed to ensure mini-map is updated
    onChangeScaleDelta(0); // reset the scale delta so can handle another click
  }, [scaleDelta]);

  const handleChangeMiniMapPosition = useCallback(() => {
    if (!stageRef.current) return;

    const stage = stageRef.current;
    // eslint-disable-next-line @typescript-eslint/no-shadow
    const scale = stage.scaleX();
    const newXPos = miniMapPosition.x * scale;
    const newYPos = miniMapPosition.y * scale;
    changePosition(newXPos, newYPos);
  }, [miniMapPosition]);

  useEffect(() => {
    handleChangeMiniMapPosition();
  }, [handleChangeMiniMapPosition]);

  useEffect(() => {
    handleChangeScaleDelta();
  }, [handleChangeScaleDelta]);

  const handleZoomStage = (event: Konva.KonvaEventObject<WheelEvent>) => {
    event.evt.preventDefault();
    if (!stageRef.current) {
      return;
    }

    const stage = stageRef.current;
    const oldScale = stage.scaleX();
    const pointerPosition = stage.getPointerPosition();
    if (!pointerPosition) {
      return;
    }

    const { x: pointerX, y: pointerY } = pointerPosition;
    const mousePointTo = {
      x: (pointerX - stage.x()) / oldScale,
      y: (pointerY - stage.y()) / oldScale,
    };

    const scaleBy = 2;
    const newScale =
      event.evt.deltaY > 0 ? oldScale * scaleBy : oldScale / scaleBy;
    if (checkIsOutOfScale(newScale)) return;

    stage.scale({ x: newScale, y: newScale });

    changePosition(
      pointerX - mousePointTo.x * newScale,
      pointerY - mousePointTo.y * newScale
    );
    onChangeScale(newScale);
  };

  const handleDragMove = () => {
    const stage = stageRef.current;
    if (!stage) {
      return;
    }
    const position = stage.getPosition();
    changePosition(position.x, position.y);
  };

  return (
    <Stage
      width={presentationSize.width}
      height={presentationSize.height}
      draggable={true}
      onWheel={handleZoomStage}
      onDragMove={handleDragMove}
      ref={stageRef}
      style={{
        cursor: 'pointer',
        borderRadius: 50,
        overflow: 'hidden',
        width: presentationSize.width,
      }}
    >
      <Layer id="sections-layer">
        <SectionsLayer
          sectionSize={historyInformation.smallImageSize}
          historyInformation={historyInformation}
          onClick={onClick}
        />
      </Layer>
    </Stage>
  );
};

export default PhotoMain;
