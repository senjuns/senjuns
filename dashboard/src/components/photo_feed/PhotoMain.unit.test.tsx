// eslint-disable-next-line import/no-extraneous-dependencies
import * as ShallowRenderer from 'react-test-renderer/shallow';
import { HistoryInformation } from '../../shared/interfaces';
import PhotoMain from './PhotoMain';

describe('<PhotoMain />', () => {
  it('renders PhotoMain', async () => {
    const renderer = ShallowRenderer.createRenderer();
    renderer.render(
      <PhotoMain
        checkIsOutOfScale={jest.fn()}
        imageSize={{ width: 1000, height: 1000 }}
        historyInformation={new HistoryInformation(2)}
        miniMapPosition={{ x: 0, y: 0 }}
        onChangePosition={jest.fn()}
        onChangeScale={jest.fn()}
        onChangeScaleDelta={jest.fn()}
        onClick={jest.fn()}
        presentationSize={{ width: 1000, height: 1000 }}
        scale={1}
        scaleDelta={1}
      />
    );

    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });
});
