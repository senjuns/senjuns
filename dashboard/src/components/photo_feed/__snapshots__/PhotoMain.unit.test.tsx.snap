// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PhotoMain /> renders PhotoMain 1`] = `
<ForwardRef
  draggable={true}
  height={1000}
  onDragMove={[Function]}
  onWheel={[Function]}
  style={
    Object {
      "borderRadius": 50,
      "cursor": "pointer",
      "overflow": "hidden",
      "width": 1000,
    }
  }
  width={1000}
>
  <Layer
    id="sections-layer"
  >
    <SectionsLayer
      historyInformation={
        HistoryInformation {
          "gridSize": Object {
            "height": 0,
            "width": 0,
          },
          "historiesGrid": Array [
            Array [],
          ],
          "isReady": false,
          "largeImageSize": Object {
            "height": 3024,
            "width": 4032,
          },
          "smallImageSize": Object {
            "height": 130,
            "width": 130,
          },
          "systemId": 2,
        }
      }
      onClick={[MockFunction]}
      sectionSize={
        Object {
          "height": 130,
          "width": 130,
        }
      }
    />
  </Layer>
</ForwardRef>
`;
